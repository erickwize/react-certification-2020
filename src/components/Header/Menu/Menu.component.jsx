import React from  'react';
import PropTypes from 'prop-types';

import {MenuBox, MiniBox } from '../styled';


const Menu = ({callBack}) => {

    const reset = () => {
        callBack()
    }

    return (
        <MenuBox onClick={reset}>
            <MiniBox space="15%"/>
            <MiniBox space="25%"/>
            <MiniBox space="25%"/>
        </MenuBox>
    );
};

Menu.propTypes = {
    callBack: PropTypes.func
}


export default Menu;