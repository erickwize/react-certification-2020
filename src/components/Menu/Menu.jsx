import React from 'react';
import { bool } from 'prop-types';
import { useAuth } from '../../providers/Auth/Auth.provider';
import styled from 'styled-components';

const StyledMenu = styled.nav`
  display: flex;
  flex-direction: column;
  justify-content: left;
  background: ${({ theme }) => theme == 'headerMenuDark' ? 'rgba(66,66,66,1)' : 'white'};
  transform: ${({ open }) => open ? 'translateX(0)' : 'translateX(-100%)'};
  height: 100vh;
  width: 30vh;
  text-align: left;
  padding: 1rem;
  position: absolute;
  top: 0;
  left: 0;
  transition: transform 0.3s ease-in-out;
  border-width: 1px;
  
  a {
    
    color: ${({ theme }) => theme == 'headerMenuDark' ? 'white' : 'black'};

    &:hover {
      background: ${({ theme }) => theme == 'headerMenuDark' ? 'rgba(77,77,77,1)' : 'rgba(245,245,245,1)'};
    }
  }
`;

const MenuLink = ({ title, linkTo, isPrivate, setOpen, isAuthenticated }) => {
  let show = true;
  if(isPrivate && !isAuthenticated) {
    show = false;
  }

  const onSetOpen = () => {
    setOpen(false);
  }

  return (
    <div>
    {show ? 
      <a href={linkTo} onClick={onSetOpen}>{title}</a>
      : null
    }
    </div>
  )
}

const Menu = ({ theme, open, setOpen }) => {
  const { authenticated } = useAuth();

  return (
    <StyledMenu open={open} theme={theme}>
      <br />
      <br />

      <MenuLink 
        title="Home" 
        linkTo="/" 
        setOpen={setOpen}
        isPrivate={false} />

      <MenuLink 
        title="Favorites" 
        linkTo="/#/favorites" 
        setOpen={setOpen}
        isPrivate={true} 
        isAuthenticated={authenticated} />

    </StyledMenu>
  )
}

Menu.propTypes = {
  open: bool.isRequired,
}

export default Menu;