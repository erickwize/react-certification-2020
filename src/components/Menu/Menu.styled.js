import styled from 'styled-components'

const StyledMenu = styled.div.attrs({
  className: 'menu',
})`
  display: flex;
  z-index: 9;
  flex-direction: column;
  justify-content: flex-start;
  background: ${({ toggle }) => (toggle ? '#2d314d' : '#FFFFFF')};
  height: 100vh;
  text-align: left;
  width: ${({ toggle }) => (toggle ? '8em' : '0em')};
  height: 100vh;
  position: fixed;
  top: 0;
  left: 0;
  transition-property: background, width;
  transition-duration: 0.1s, 0.1s;
  transition-timing-function: ease, ease;
  @media only screen and (max-width: 600px) {
    width: ${({ toggle }) => (toggle ? '8em' : '0')};
  }
  a {
    margin-left: 1em;
    margin-top: 2em;
    font-weight: bold;
    opacity: ${({ toggle }) => (toggle ? '1' : '0')};
    transition: ${({ toggle }) => (toggle ? 'opacity 0.1s ease' : 'opacity 0.1s ease')};
    &:last-child {
      margin-top: 0em;
    }
  }
`
export default StyledMenu
