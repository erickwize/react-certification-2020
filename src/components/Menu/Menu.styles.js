import { motion } from 'framer-motion';
import styled from 'styled-components';

export const NavContainer =  styled(motion.nav)`
  position:fixed;
  height: 100%;
  width: 10%;
  z-index: 3;

  @media(min-width:280px) and (max-width:414px){
    width: 100%;
  }

  @media(min-width:415px) and (max-width:1024px){
    width: 35%;
  }

  svg{
    fill: ${props => props.theme.header.svgColor};
  }
`;

export const StyledButton = styled(motion.button)`
  background-color: transparent;
  border: none;
  border-radius: 50%;
  padding: 0px;
  height: 50px;
  width: 50px;
  margin: 15px 0 0 15px;
  
  &:focus{
    outline: none;
  }

  &:hover{
    cursor: pointer;
  }
`;