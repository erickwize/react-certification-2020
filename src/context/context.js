// import Dexie from 'dexie'
import React, { useEffect, createContext, useReducer, memo } from 'react'
import { useFetch } from '../utils/hooks/useFetch'

export const Context = createContext(null)

export const ContextReducer = (state, action) => {
  switch (action.type) {
    case 'TOGGLE_THEME':
      return {
        ...state,
        theme: !state.theme,
      }
    case 'OPEN_MENU':
      return {
        ...state,
        menu: action.payload,
      }
    case 'GET_DATA':
      return {
        ...state,
        data: action.payload,
      }
    case 'SET_VIEW':
      return {
        ...state,
        view: action.payload,
      }
    case 'SET_TARGET':
      return {
        ...state,
        target: action.payload,
      }
    default:
      return state
  }
}

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data, isLoading } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])

  return isLoading ? (
    <div> Loading...</div>
  ) : (
    <Context.Provider value={{ state, dispatch }}> {children} </Context.Provider>
  )
}
export default memo(ContextProvider)
