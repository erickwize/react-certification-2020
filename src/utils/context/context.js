import { createContext } from 'react';

export const Context = createContext({
  queryToSearch: null,
  setQueryFromSearchInput: null,
});
