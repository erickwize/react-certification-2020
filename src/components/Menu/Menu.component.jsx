import React from 'react';
import { MenuSVG } from '../../svg/Menu';
import { MenuList } from './MenuList/MenuList.component';
import { useGlobal } from '../../providers/Global.provider';
import { NavContainer, StyledButton } from './Menu.styles';


const variants = {
  open: (custom) => ({
    backgroundColor: `${custom.backgroundColorOpen}`,
    clipPath: `circle(${1000 * 2 + 200}px at 40px 40px)`,
    transition: {
      type: "spring",
      stiffness: 20,
      restDelta: 2
    }
  }),
  closed: (custom) => ({
    backgroundColor: `${custom.backgroundColorClosed}`,
    clipPath: "circle(30px at 40px 40px)",
    transition: {
      delay: 0,
      type: "spring",
      stiffness: 400,
      damping: 40
    }
  })
};

export const Menu = () => {
  const { state, dispatch } = useGlobal();
  const { theme, isMenuOpen } = state;
  return <NavContainer animate={isMenuOpen ? "open" : "closed"} variants={variants} theme={theme} custom={theme.header.menu}>
    <StyledButton onClick={() => dispatch({type:'menu'})}>
      <MenuSVG/>
      </StyledButton>
    <MenuList/>
  </NavContainer>
}