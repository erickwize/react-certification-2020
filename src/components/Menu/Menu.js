import React, { useContext } from 'react'
import { Context } from '../../context/context'
import StyledMenu from './Menu.styled'

const Menu = React.memo(() => {
  const { state, dispatch } = useContext(Context)
  const changeToHome = (e) => {
    e.preventDefault()
    dispatch({ type: 'SET_VIEW', payload: 'home' })
    dispatch({ type: 'OPEN_MENU', payload: false })
    dispatch({ type: 'SET_TARGET', payload: 'Wizeline' })
  }

  const changeToFavorites = (e) => {
    e.preventDefault()
    dispatch({ type: 'SET_VIEW', payload: 'favorites' })
    dispatch({ type: 'OPEN_MENU', payload: false })
    dispatch({ type: 'SET_TARGET', payload: '' })
  }

  return (
    <StyledMenu toggle={state.menu}>
      <a href="/" onClick={changeToHome}>
        Home
      </a>
      <a href="/" onClick={changeToFavorites}>
        Favorites
      </a>
    </StyledMenu>
  )
})
export default Menu
