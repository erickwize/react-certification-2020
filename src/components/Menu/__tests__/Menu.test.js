import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import renderer from 'react-test-renderer'
import { fireEvent } from '@testing-library/svelte'
import { render } from '@testing-library/react'
import { Context, initialState } from '../../../context/context'
import Menu from '../Menu'

const state = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
}
describe('Testing Menu component', () => {
  test('Renders Menu without crashing', async () => {
    const rendered = renderer.create(
      <Context.Provider value={{ state }}>
        <Menu />
      </Context.Provider>
    )
    expect(rendered).toMatchSnapshot()
  })
  test('Home view changed when button menu clicked', async () => {
    const { getByText } = render(
      <Context.Provider value={{ state }}>
        <Menu />
      </Context.Provider>
    )
    const linkHome = getByText('Home')

    await fireEvent(
      linkHome,
      new MouseEvent('click', { type: 'SET_VIEW', payload: 'home' })
    )
    expect(initialState.view).toEqual('home')
  })
})
