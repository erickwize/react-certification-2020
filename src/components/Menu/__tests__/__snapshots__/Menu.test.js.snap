// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Menu component Renders Menu without crashing 1`] = `
<div
  className="sc-bdvvaa bYAoZE menu"
>
  <a
    href="/"
    onClick={[Function]}
  >
    Home
  </a>
  <a
    href="/"
    onClick={[Function]}
  >
    Favorites
  </a>
</div>
`;
